jme3-core/src/main/java/com/jme3/scene/control/LightControl.java

```diff
@@ -127,49 +128,70 @@ protected void controlUpdate(float tpf) {
+        final Vector3f worldDirection = vars.vect2;
+        spatial.getWorldRotation().mult(Vector3f.UNIT_Z, worldDirection).negateLocal();
 
         if (light instanceof PointLight) {
             ((PointLight) light).setPosition(worldTranslation);
-            return;
-        }
-
-        final TempVars vars = TempVars.get();
-        final Vector3f vec = vars.vect1;
-
-        if (light instanceof DirectionalLight) {
-            ((DirectionalLight) light).setDirection(vec.set(worldTranslation).multLocal(-1.0f));
-        }
-
-        if (light instanceof SpotLight) {
+        } else if (light instanceof DirectionalLight) {
+            ((DirectionalLight) light).setDirection(worldDirection);
+        } else if (light instanceof SpotLight) {
             final SpotLight spotLight = (SpotLight) light;
             spotLight.setPosition(worldTranslation);
-            spotLight.setDirection(spatial.getWorldRotation().multLocal(vec.set(Vector3f.UNIT_Y).multLocal(-1)));
+            spotLight.setDirection(worldDirection);
         }
-
         vars.release();
     }
 
+    /**
+     * Sets the spatial to adopt the light's world transformations.
+     *
+     * @author Markil 3
+     */
     private void lightToSpatial(Light light) {
         TempVars vars = TempVars.get();
-        if (light instanceof PointLight) {
+        boolean rot = false, trans = false;
 
+        if (light instanceof PointLight) {
             PointLight pLight = (PointLight) light;
-
-            Vector3f vecDiff = vars.vect1.set(pLight.getPosition()).subtractLocal(spatial.getWorldTranslation());
-            spatial.setLocalTranslation(vecDiff.addLocal(spatial.getLocalTranslation()));
+            vars.vect1.set(pLight.getPosition());
+            trans = true;
+        } else if (light instanceof DirectionalLight) {
+            DirectionalLight dLight = (DirectionalLight) light;
+            vars.vect2.set(dLight.getDirection()).negateLocal();
+            rot = true;
+        } else if (light instanceof SpotLight) {
+            SpotLight sLight = (SpotLight) light;
+            vars.vect1.set(sLight.getPosition());
+            vars.vect2.set(sLight.getDirection()).negateLocal();
+            trans = rot = true;
+        }
+        if (spatial.getParent() != null) {
+            spatial.getParent().getLocalToWorldMatrix(vars.tempMat4).invertLocal();
+            vars.tempMat4.rotateVect(vars.vect1);
+            vars.tempMat4.translateVect(vars.vect1);
+            vars.tempMat4.rotateVect(vars.vect2);
         }
 
-        if (light instanceof DirectionalLight) {
-            DirectionalLight dLight = (DirectionalLight) light;
-            vars.vect1.set(dLight.getDirection()).multLocal(-1.0f);
-            Vector3f vecDiff = vars.vect1.subtractLocal(spatial.getWorldTranslation());
-            spatial.setLocalTranslation(vecDiff.addLocal(spatial.getLocalTranslation()));
+        if (rot) {
+            vars.quat1.lookAt(vars.vect2, Vector3f.UNIT_Y);
```

Comment:
If `vect2` is (0, 1, 0) or (0, -1, 0) then `lookAt()` will produce an invalid rotation.

Reply:
That is a bit odd. I've tested TestLightControl2Spotlight with the spotlight set to multiples of UNIT_Y, but I'm not seeing anything odd. The spotlight does point in the right direction.
That said, if I try replacing the red spotlight model with something a little more detailed, such as Models/MonkeyHead/MonkeyHead.mesh.xml, rotating the spotlight along the X axis reveals some odd behavior where the model does not like to be upside down, and will flip 180 degrees on the local vertical axis so that the crown of the head points in a y direction greater than 0. This flip will happen at the aforementioned vectors. I suspect that this has to do with the limitations of using just directions for the lights rather than a full-blown rotation.
[Demonstration of the problem. Note that the back of the monkey head always points towards the light, even if the monkey did get rotated.](https://github.com/jMonkeyEngine/jmonkeyengine/files/6103391/monkeyrotdemo.zip)
At this point, though, I am a little stumped as to how to go about fixing this. I've had limited success negating the second parameter of lookAt at certain points, but it honestly doesn't feel like the right way to go about it.

Reply:
I think it's reasonable for a `LightControl` in `LightToSpatial` mode to flip a model that doesn't have rotational symmetry around its Z axis. That's a consequence of (non-point) lights having a direction but not a full-blown rotation. It's also a feature worth documenting, in my opinion.

What's puzzling me is that the invalid rotations don't cause visible distortion of the model, even though I can demonstrate distortion using `lightNode.getWorldTransform().transformVector()`. (Perhaps the renderer normalizes the quaternion at some point?) To avoid invalid rotations, I suggest normalizing `quat1` by adding `vars.quat1.normalizeLocal();` immediately after the `lookAt()`.

By the way, this code would be much more readable if you set up aliases for `vect1` and `vect2` in `lightToSpatial()`, as has been done in `spatialToLight()`. Similarly, `rot` and `trans` should have more descriptive names: perhaps `rotateSpatial` and `translateSpatial`.

Reply:
I just had an idea for avoiding sudden model flips. Instead of using `Vector3f.UNIT_Y` as the "up" direction in `lookAt()` you might use the model's current local "up" direction, that is, `spatial.getLocalRotation().mult(Vector3f.UNIT_Y, vars.vect3)`.

Reply:
I tried your last fix, but it doesn't seem to reset rotational changes appropriately. When I replace the cone model with Models/MonkeyHead/MonkeyHead.mesh.xml, the model goes through the X, Y, and Z rotations quite well. However, after it rotates on the random axis, it doesn't reset when it goes back to X, resulting in it rotating at an odd angle.

Reply:
Thanks for testing my idea. So either we have a consistent "up" direction or we avoid sudden flips. While it would be nice to have both features, I don't see a way to achieve that.